### PR TITLE
Convert App to use hooks

### PR DESCRIPTION
**Pull request by harrygr**
This allows better encapsulation of logic for a component that's getting pretty unwieldy.
More work can still be done to improve this but it's still much better

### CODE REVIEW

**Comment by harrygr** [comment]
@Jarrydcrawford you reckon I can merge this bad boi?

**Comment by Jarrydcrawford** [comment]
I shall looksie after lunch fren

File: src/App.tsx [hunk 1]
```diff
@@ -38,87 +40,126 @@ interface Props extends RouteComponentProps {
...
+    };
+
+    const setSplash = (splash: {
+      children: React.ReactNode;
+      buttonText: string;
+    }) => {
+      dispatch(SetSplash(splash));
+    };
+
+    const dismissSplash = () => dispatch(DismissSplash());
```
Comment (harrygr) [inline src/App.tsx]: We could encapsulate the splash behaviour into a custom hook if needed, although maybe it'll be better just to get rid of a splash and explore a simpler solution

File: src/App.tsx [hunk 2]
```diff
@@ -38,87 +40,126 @@ interface Props extends RouteComponentProps {
...
+    case 'SetSplash':
+      return { ...state, splash: action.splash };
+    case 'DismissSplash':
+      return { ...state, splash: undefined };
+    default:
+      throw new Error(`Unsupported AppAction ${action}`);
   }
+};
+
+export const App = withRouter(
```
Comment (harrygr) [inline src/App.tsx]: There's no `useRouter` hook yet 😢

File: src/App.tsx [hunk 3]
```diff
@@ -38,87 +40,126 @@ interface Props extends RouteComponentProps {
...
+const appReducer = (state: State, action: AppAction): State => {
+  switch (action.type) {
+    case 'SetUser':
+      return { ...state, user: action.user };
+    case 'SetSplash':
+      return { ...state, splash: action.splash };
+    case 'DismissSplash':
+      return { ...state, splash: undefined };
+    default:
+      throw new Error(`Unsupported AppAction ${action}`);
```
Comment (Jarrydcrawford) [inline src/App.tsx]: I wonder why in JS you can instantiate the constructor or call it as a function, only for the Error though.

File: src/App.tsx [hunk 4]
```diff
@@ -38,87 +40,126 @@ interface Props extends RouteComponentProps {
...
-
-  public render(): JSX.Element {
-    const isLoggedIn = this.state.user.isSome();
+    const [alertState, alert] = useAlert();
+
+    /**
+     * Check that the token in storage is actually valid
+     * for making requests. If it isn't then clear it out.
+     */
+    const checkAuth = (): Promise<any> => {
```
Comment (Jarrydcrawford) [inline src/App.tsx]: Not really part of this change, but should be `Promise<void>` 🤷‍♂️

File: src/App.tsx (hunk 3 above)
Comment (harrygr) [reply src/App.tsx]: yeah the `Error` is a strange one. I guess `Error` behaves as a factory function when called directly. 

Calling `Date` directly for example just returns a string.
